Disable starting builds of a particular type when you don't have access to the builds subresource

csrwng opened the pull request:
Fixes #4017 
Bug 1246426

csrwng commented:
@bparees @liggitt ptal

smarterclayton commented on pkg/build/admission/admission.go:
```diff
@@ -43,21 +43,57 @@ func NewBuildByStrategy(client client.Interface) admission.Interface {
 	buildsResource       = "builds"
 	buildConfigsResource = "buildconfigs"
+
+	cloneSubresource       = "clone"
```
are these constants really helping?  used in one place...

csrwng replied:
they need a better place to live ... I'll move them. We need to use them wherever we reference these resources like the boostrap policy

csrwng replied:
got rid of the subresource constants per @liggitt's comment below. So far it seems we're not using constants for anything else in master.go and boostrappolicy/policy.go, so I'll leave the 2 existing ones there since they're used at least twice in this file.

bparees commented:
lgtm

liggitt commented on pkg/build/admission/admission.go:
```diff
@@ -43,21 +43,57 @@ func NewBuildByStrategy(client client.Interface) admission.Interface {
-	case *buildapi.Build:
-		err = a.checkBuildAuthorization(obj, attr)
-	case *buildapi.BuildConfig:
-		err = a.checkBuildConfigAuthorization(obj, attr)
```
do we not just want a `*buildapi.BuildRequest` case with a `checkBuildRequestAuthorization()` func that does the build lookup and then calls `checkBuildAuthorization`?

liggitt replied:
I don't even know that I'd check the subresource type... if you're creating a BuildRequest, we want to make sure you're allowed...

csrwng replied:
> do we not just want a *buildapi.BuildRequest case with a checkBuildRequestAuthorization() func that does the build lookup and then calls checkBuildAuthorization?

The problem is that the meaning of BuildRequest.Name changes depending on whether it's sent with an bc/instantiate or a build/clone subresource. It's a bc name in the first case, build name in the second case.

liggitt replied:
so pass in `attr.getResource()` to checkBuildRequestAuthorization()

csrwng replied:
true, will refactor

liggitt commented on pkg/build/admission/admission.go:
```diff
@@ -107,6 +107,24 @@ func (a *buildByStrategy) checkBuildConfigAuthorization(buildConfig *buildapi.Bu
+			return err
+		}
+		return a.checkBuildConfigAuthorization(build, attr)
+	}
```
if we get a build request from a type we don't recognize, I think we should error... it probably means something new got created that we need to handle here (or at least explicitly not handle)

csrwng replied:
fixed

liggitt replied:
I actually meant a `default: return forbidden` case here, not above (I'm not familiar enough with the build APIs to know if we submit any other types)

csrwng replied:
I would never get that here because if it's not one of the expected resources, we return nil at the beginning:
https://github.com/openshift/origin/pull/4032/files#diff-3eae75d1c69256688c5920b3899663bfL49
Afaik there is no other type of object that we submit for builds other than BuildConfig, Build, and BuildRequest, so having the check at the top is ok. If we want to allow other type of subresource objects in the future, we'd have to explicitly change this code to do that.

liggitt replied:
ah. I'd still like the `default:` here, so this method is correct without making assumptions about how it's called :)

csrwng replied:
updated

liggitt commented:
add clone and instantiate calls to the existing integration test to make sure admission is checking the correct user for builds/clone and buildConfigs/instantiate requests and denies appropriately

csrwng commented:
completely forgot about the int test :)

csrwng commented:
added integration tests, changed the way we generate build names for build clone to use the ObjectMeta.GenerateName field (integration test was failing when cloning builds in quick succession)

liggitt commented on pkg/build/generator/generator.go:
```diff
@@ -515,8 +513,8 @@ func generateBuildFromBuild(build *buildapi.Build) *buildapi.Build {
 		ObjectMeta: kapi.ObjectMeta{
-			Name:   getNextBuildNameFromBuild(buildCopy),
-			Labels: buildCopy.ObjectMeta.Labels,
+			GenerateName: buildCopy.Name + "-",
```
if you clone a clone, this will be the result, right?
```
build-1-asldkaj
build-1-asldkaj-qwueqw
build-1-asldkaj-qwueqw-xsasld
```

what would the old method do with a clone of a clone?

csrwng replied:
hmm, it would replace the last part ("-" separated) if it was made up of numbers.

csrwng replied:
need to give it a little bit more thought :)

liggitt replied:
yeah, I think that was to avoid the clone-of-a-clone issue

csrwng commented:
ok, second attempt to fix name generation

liggitt commented on pkg/build/generator/generator.go:
```diff
@@ -532,7 +532,9 @@ func getNextBuildNameFromBuild(build *buildapi.Build) string {
 	}
-	return fmt.Sprintf("%s-%d", buildName, int32(util.Now().Unix()))
+	suffix := fmt.Sprintf("%v", util.Now().UnixNano())
+	suffix = suffix[len(suffix)-10:]
```
make sure the length is at least 10

csrwng replied:
this is when  I wish go had a max function :-)

liggitt replied:
you never know when time is going to run backward

csrwng replied:
fixed

liggitt commented on pkg/build/generator/generator.go:
```diff
@@ -532,7 +532,11 @@ func getNextBuildNameFromBuild(build *buildapi.Build) string {
 		buildName = strings.Join(nameElems[:len(nameElems)-1], "-")
 	}
-	return fmt.Sprintf("%s-%d", buildName, int32(util.Now().Unix()))
+	suffix := fmt.Sprintf("%v", util.Now().UnixNano())
```
what happens if this makes us exceed the max build name length? I know the old method (epoch) resulted in something at least 10 chars long, so we're no worse off, just wondering what happens

csrwng replied:
I think it just fails... will fix

csrwng replied:
Using namer now

csrwng commented:
@liggitt - do you have any additional comments on this PR ?

smarterclayton commented:
I want to take a quick glance tonight but won't hold it up tomorrow if I
don't get to it

On Aug 11, 2015, at 11:30 AM, Cesar Wong <notifications@github.com> wrote:

@liggitt <https://github.com/liggitt> - do you have any additional comments
on this PR ?

—
Reply to this email directly or view it on GitHub
<https://github.com/openshift/origin/pull/4032#issuecomment-129930650>.

liggitt commented:
the actual changes LGTM, the changes to the integration test were more disruptive than I was expecting, so I'm trying to re-read them all to make sure we're testing what we think we're testing

csrwng commented:
[test]

liggitt commented on test/integration/build_admission_test.go:
```diff
@@ -13,165 +13,148 @@ import (
 	testutil "github.com/openshift/origin/test/util"
 )
 
 func TestPolicyBasedRestrictionOfBuildStrategies(t *testing.T) {
```
maybe combine the create/clone tests to save setup/test time:
1. verify create succeeds and save result
2. verify clone of saved succeeds
3. remove perms
4. verify create fails
5. verify clone of saved build fails

liggitt commented on test/integration/build_admission_test.go:
```diff
@@ -13,165 +13,148 @@ import (
-	}
-	if _, err := clusterRoleInterface.Update(role); err != nil {
-		t.Errorf("unexpected error: %v", err)
+	clients := map[string]*client.Client{"admin": projectAdminClient, "editor": projectEditorClient}
```
same for buildconfig/instantiate? combine to save time?

liggitt commented:
comment on combining tests, then LGTM

csrwng commented:
updated integration test

liggitt commented:
LGTM, [merge]
